adobe/zookeeper-operator
# Add the multiAddress.enabled cluster option present in zk >3.6.0

### Change log description
- Enable the multiaddreses.Enabled JVM flag in the zkserver.sh
- parse inputted extra addresses and add them to the zoo config
_(2-3 concise points about the changes in this PR. When committing this PR, the committer is expected to copy the content of this section to the merge description box)_

### Purpose of the change
In zookeeper > 3.6.0 you can specify multiple addresses for a singular zookeeper server. [here](https://zookeeper.apache.org/doc/r3.9.1/zookeeperAdmin.html#sc_clusterOptions) is the description of that cluster option. This update allows you to leverage that feature through the operator. This can also be useful if you want to expose individual zk servers through external addresses.
_(e.g., Fixes #666, Closes #1234)_

### What the code does
- Add funcitionality to parse a list of zk server addresses in additionalConfig when creating the config map
- in the zookeeper image, parse those additional addresses and properly use them in the zk config
  - The operator currently writes server addresses on write config and node registration. When one of these events happen we add an extra step that:
    - parses the new addServerAddresses.txt created by the config map for addresses listed for the current server ID
    - if there is a "|" it formats each address using the zkconfig function, otherwise it adds the singular address
    - it also removes the client port that is appended by the zkconfig function for all addresses except the final one in the list. This is to follow formatting for the multiAddress feature as the client port should only be present once per list of servers

_(Detailed description of the code changes)_

### How to verify it
1. Deploy the operator via a helm chart
2. under additionalConfig, add extra server addresses associated with zk servers (In our case it is the addresses for the dedicated VIPs that expose each server)
    ```
    additionalConfig: {
      server.1: query-service-stg-zookeeper-0.ethos11-stage-nld2.ethos.adobe.net,
      server.2: query-service-stg-zookeeper-1.ethos11-stage-nld2.ethos.adobe.net,
      server.3: query-service-stg-zookeeper-2.ethos11-stage-nld2.ethos.adobe.net,
      server.4: query-service-stg-zookeeper-3.ethos11-stage-nld2.ethos.adobe.net,
      server.5: query-service-stg-zookeeper-4.ethos11-stage-nld2.ethos.adobe.net
    }
    ```
  - You can add addresses for some, all or no servers. You can also add multiple addresses for the same server in a pipe separated list.
    ```
     additionalConfig: {
      server.1: query-service-stg-zookeeper-0.ethos11-stage-nld2.ethos.adobe.net,
      server.3: query-service-stg-zookeeper-2.ethos11-stage-nld2.ethos.adobe.net,
      server.5: query-service-stg-zookeeper-4.ethos11-stage-nld2.ethos.adobe.net
    }
    ```

    ```
     additionalConfig: {
      server.1: query-service-stg-zookeeper-0.ethos11-stage-nld2.ethos.adobe.net|query-service-stg-zookeeper.ethos11-stage-nld2.ethos.adobe.net,
      server.2: query-service-stg-zookeeper-1.ethos11-stage-nld2.ethos.adobe.net,
      server.3: query-service-stg-zookeeper-2.ethos11-stage-nld2.ethos.adobe.net|www.google.com,
      server.4: query-service-stg-zookeeper-3.ethos11-stage-nld2.ethos.adobe.net,
      server.5: query-service-stg-zookeeper-4.ethos11-stage-nld2.ethos.adobe.net
     }
    ```
3.  Validate correct address list is displayed through k8s portforwarding
     ```
      kubectl get pods
      kubectl port-forward app-zookeeper-0 2181:2181
     ```
    ```
    castaffo@castaffo-mac zookeeper-operator % echo conf | nc 127.0.0.1 2181
    ...
    membership: 
    server.1=app-zookeeper-0.app-zookeeper-headless.ns-team-experience-platform--query-service-zookeeper--6584bc25.svc.cluster.local:2888:3888|query-service-stg-zookeeper-0.ethos11-stage-nld2.ethos.adobe.net:2888:3888|query-service-stg-zookeeper.ethos11-stage-nld2.ethos.adobe.net:2888:3888:participant;0.0.0.0:2181
    server.2=app-zookeeper-1.app-zookeeper-headless.ns-team-experience-platform--query-service-zookeeper--6584bc25.svc.cluster.local:2888:3888|query-service-stg-zookeeper-1.ethos11-stage-nld2.ethos.adobe.net:2888:3888:participant;0.0.0.0:2181
    server.3=app-zookeeper-2.app-zookeeper-headless.ns-team-experience-platform--query-service-zookeeper--6584bc25.svc.cluster.local:2888:3888|query-service-stg-zookeeper-2.ethos11-stage-nld2.ethos.adobe.net:2888:3888|www.google.com:2888:3888:participant;0.0.0.0:2181
    server.4=app-zookeeper-3.app-zookeeper-headless.ns-team-experience-platform--query-service-zookeeper--6584bc25.svc.cluster.local:2888:3888|query-service-stg-zookeeper-3.ethos11-stage-nld2.ethos.adobe.net:2888:3888:participant;0.0.0.0:2181
    server.5=app-zookeeper-4.app-zookeeper-headless.ns-team-experience-platform--query-service-zookeeper--6584bc25.svc.cluster.local:2888:3888|query-service-stg-zookeeper-4.ethos11-stage-nld2.ethos.adobe.net:2888:3888:participant;0.0.0.0:2181

  ```
